wdl-format/src/metadata.rs

```diff
@@ -0,0 +1,333 @@
+//! A module for formatting metadata sections (meta and parameter_meta).
+
+use wdl_ast::v1::LiteralNull;
+use wdl_ast::v1::MetadataArray;
+use wdl_ast::v1::MetadataObject;
+use wdl_ast::v1::MetadataObjectItem;
+use wdl_ast::v1::MetadataSection;
+use wdl_ast::v1::MetadataValue;
+use wdl_ast::v1::ParameterMetadataSection;
+use wdl_ast::AstNode;
+use wdl_ast::AstToken;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+
+use super::comments::format_inline_comment;
+use super::comments::format_preceding_comments;
+use super::first_child_of_kind;
+use super::formatter::SPACE;
+use super::Formattable;
+use super::Formatter;
+use super::NEWLINE;
+
+impl Formattable for LiteralNull {
+    fn format<T: std::fmt::Write>(
+        &self,
+        writer: &mut T,
+        _state: &mut Formatter,
+    ) -> std::fmt::Result {
+        write!(writer, "{}", self.syntax())
+    }
+}
+
+impl Formattable for MetadataObject {
+    fn format<T: std::fmt::Write>(
+        &self,
+        writer: &mut T,
+        formatter: &mut Formatter,
+    ) -> std::fmt::Result {
+        format_preceding_comments(
+            &SyntaxElement::from(self.syntax().clone()),
+            writer,
+            formatter,
+            false,
+        )?;
+
+        let open_brace = first_child_of_kind(self.syntax(), SyntaxKind::OpenBrace);
+        format_preceding_comments(&open_brace, writer, formatter, true)?;
+        // Open braces should ignore the "+1 rule" followed by other interrupted
+        // elements.
+        if formatter.interrupted() {
+            formatter.reset_interrupted();
+            formatter.indent(writer)?;
+        }
+        write!(writer, "{}", open_brace)?;
+        format_inline_comment(&open_brace, writer, formatter, false)?;
+
+        formatter.increment_indent();
+
+        let mut commas = self
```

Comment:
What's the point of this whole comma checking business? Aren't we just going to
output a comma to every line regardless of what the source has?

Reply:
This is required by how the comment checking works. Every comment in the file is "owned" by a `SyntaxElement`. A metadata `Comma` may own one or more comments, requiring us to iterate through each comma and check it for comments. We can't just ignore commas as trivia and insert them ourselves, because that would cause us to "lose" comments in the output.